### DISCUSSION TITLE
Reroute incoming http request if rerouting target path exists

### DISCUSSION THREAD

**Discussion by HerrDerb** [quarkusio/quarkus#44444] [Q&A]
For a little api versioning feature, I would like to "override" a couple of endpoint in versioned controller classes (Simply added a version at the begin of the base path). Those controller may only define endpoints which are relevant for this version, e.g. breaking change on the next version. 
I would like to reroute an incoming request in a `ContainerRequestFilter` to the corresponding versioned controller if necessary. For this, I am using the version which I receive as a header, look out for a relevant versioned controller and if one found check if it contains an overridden version of the current request path. The rerouting then is done by simply adding the version to the begin of the request path. I did understand that a rerouting needs to be done in a `PreMatching` filter.

Now I have the problem that I could not found a reasonable way to make the matching. I see two possibilities for which I both miss a piece:
1. Preferred: Match the `@Path` expression of the versioned controller to the current request path -> I could not find a util or class that conveniently allows me to do that. Does this exist?
2. Resolve the original endpoint of the incoming request and use it's `@Path` annotation to find a matching endpoint on my versioned controllers. With this approach, I was not able to find a way to resolve an incoming request to the target endpoint (method) in a `Prematching` `ContainerRequestFilter`. Is there a util or a way to resolve an incoming request? Maybe somehow with a `RequestMapper` which I could not find out on how to access if possible at all.

Is it possible to resolve my problem with on of the both described ways?

# Example
If have this default controller
```java
@Path("/api")
class ItemController  {

    @GET
    @Path("/items/")
    public List<Items> getItems() {
        return getItems();
    }

    @Post
    @Path("/items/")
    public void createItem(Item item) {
         createItem(item);
    }
}
```
and this versioned controller which handles request up to 10.0 which only "overrides" the GET endpoint

```java
@Path("/v10.0/api")
class V10ItemController implements VersionedController {

    @Overriden
    public Version getTargetingMaxVersion() {
        return new Version(10, 0);
    }

    @GET
    @Path("/items/")
    public Response getItems() {
        return Response.ok().build();
    }
}
```

On an incoming request like:
Example incoming request:
```http
GET http://localhost:8080/api/items/
Headers: {'X-APP-Version': '9.0'}
```

I select the matching versioned controller and want to reroute to the overridden endpoint if exists. In this case, the rerouted request would look like:
```http
GET http://localhost:8080/v10.0/api/items/
Headers: {'X-APP-Version': '9.0'}
```

While selecting the versioned controller is not a problem, I have the problem (described above) to verify if the versioned controller contains an overridden endpoint for the incoming request.

**Comment by HerrDerb** [comment]
Maybe  @FroMage @stuartwdouglas or @geoand might know where to look?

**Reply by FroMage** [reply]
Could you give us some examples of paths, controllers and methods and how you'd like to reroute things? It'd make things more concrete for us to make sense of what you're trying to do.

**Reply by HerrDerb** [reply]
@FroMage Sure, I am happy to provide more details. I added a example to the original question. I hope this makes my question more clear.

**Reply by FroMage** [reply]
OK thanks.

I don't think you can obtain the original matching endpoint in a pre-match filter, because this is before we do any matching. As a result, you'd have to handle the resolving yourself, which can be straightforward for simple cases.

I have implemented versioned REST APIs in the past, in at least two different ways, and one option you could use is inheritance, where your default endpoint is up to date (latest version) and you use subclasses for older versions:

```java
@Path("/api")
class ItemController  {

    @GET
    @Path("/items/")
    public List<Items> getItems() {
        return getItems();
    }

    @Post
    @Path("/items/")
    public void createItem(Item item) {
         createItem(item);
    }
}

@Path("/v10.0/api")
class V10ItemController extends ItemController {

   @Override
    @GET
    @Path("/items/")
    public Response getItems() {
        return Response.ok().build();
    }
}
```

So in this case your pre-match filter will redirect everything < 10 to `/v10.0/api` and `GET /v10.0/api/items/` will be V10, while `POST /v10.0/api/items/` will be latest version.

Or, do it the other way around, where the latest version is an override of previous versions:

```java
@Path("/api")
class ItemController  extends V10ItemController {

   @Override
    @GET
    @Path("/items/")
    public List<Items> getItems() {
        return getItems();
    }

    @Post
    @Path("/items/")
    public void createItem(Item item) {
         createItem(item);
    }
}

@Path("/v10.0/api")
class V10ItemController {

    @GET
    @Path("/items/")
    public Response getItems() {
        return Response.ok().build();
    }
}
```

Same pre-match filter will redirect everything < 10 to `/v10.0/api` and `GET /v10.0/api/items/` will be V10, while `POST /v10.0/api/items/` will not exist.

Other options include:

- Pre-match filter that doesn't do re-routing but redirects
- Making your clients use explicit versioned APIs: no more `/api` to mean "latest" but `/api/v12.1` explicitly so no need for rerouting
- Handling version differences in the actual endpoints. Here you could have two different outcomes inside `getItems()` depending on the requested version, with a single endpoint handling every API version.
- Another version of the same idea is to have a single endpoint class, and delegate every business logic to versioned service beans: `switch(versionHeader()){ case V10: return v10Service.getItems(); case V12: return v12Service.getItems();}`
- Custom serialisation of the `Item` class depending on the requested version. Using Jackson features. If this is just about including some fields conditionally, or serialising them differently.

Also, there's a more interesting option if you want to support negociation of API version between clients and servers, but this is when you want clients to support multiple server versions until API X, and servers to support multiple client versions until API Y and you have to find a common version to communicate with.

Does this help?

**Reply by HerrDerb** [reply]
Hm I was thinking about extending the "latest" controller but then I might miss other method overrides of a later versions. 
My goal would be a mechanism of which the client does not need to know about. Also I would like to keep the "latest" code clean from that. In other work, the whole backwards compatibility is decoupled.
One doable way would be to match the url to the `@Path` expression using a regex, but the framework somehow needs to do this too, so I rather would rely on that code if possible
Maybe you can help to point out where the actual mapping of the request to the controller method  in Quarkus\Jax-rs happens during runtime?

**Reply by HerrDerb** [reply]
I got a working version by converting the path expressions to a regex like:
```
    public static Pattern generateRegexFromPath(String pathAnnotation) {
        // Escape special regex characters and replace path parameters with regex patterns
        String regex = pathAnnotation.replaceAll("\\{[^/]+}", "[^/]+");
        // Add start and end anchors and ignore leading /
        regex = "^/?" + regex + "$";
        return Pattern.compile(regex);
    }
```
With the prematching filter I've got now, I can route a request to the versioned controller that handles a breaking changes for the incoming version.

Lets say the incoming request comes from app with version 8.0. After Version 10.2 I have a breaking change on Path `/abc`.
I start checking all versioned controller for Path `/abc` starting at version 8.0, I will get a match with 10.2 and simply add `v10.2` to the path and reroute. 

This way the client does not need to know about and the the latest code neither. I only have to define a versioned controller with an "override" of the endpoint that has breaking changes for the next version.

**Reply by FroMage** [reply]
OK, great :)

**Reply by HerrDerb** [reply]
@FroMage so just to be sure 🙂 there is no way, within the scope of a pre matching filter, on how I could use any existing mechanism in quarkus to avoid the custom regex matching?

**Reply by FroMage** [reply]
Not that I know of, no.

**Reply by HerrDerb** [reply]
@FroMage Thanks for you time and inputs 👍🙂

**Comment by HerrDerb** [comment]
Found a working solution in the scope of a prematching filter
